### ISSUE TITLE
Problems with Latest actualizations 0.9.2.2 

### ISSUE THREAD

**Issue by joancf** [gephi/gephi#1854]
## the graph.getEdge(node,node) function returns null 
The same graph before and after the actualization performs differently under the multmode-networks plugin. 
Graphs that before the actualization perform well, return a zero edges graph after the actualization.
The problem seems to be that the  graph.getEdge(node,node) function always returns null after the actualization.

## Steps to Reproduce

2. Install the multimode plugin
2. Load the attached file
[barcelona_user_words_tp1.graphml.zip](https://github.com/gephi/gephi/files/1500972/barcelona_user_words_tp1.graphml.zip)
3. do the multimode on type (word-user user-word) remove edges/nodes
4.A 0 edges graph is produced after the module ends

## Code and Debug problems
I could not get the last code version of the module , the module's link is broken and it is impossible to know which of the hundred forks of the gephi plugins is the one used to generate de distribution code...

I can debug the 0.9.2-SNAPSHOT version of gephi but I don't know how to  include the latest actualizations  using Maven (**I need some help on that!!!**)  

## My Environment

* Version used: Gephi 0.9.2
* Java version:  oracle-8
* Operating System: Linux

**Comment by eduramiba** [comment]
Hi,
This method is part of Graphstore, which has not had a release since August, and has multiple tests covering that. https://github.com/gephi/graphstore/releases

I guess this pull request https://github.com/gephi/gephi-plugins/pull/131 introduced a bug, probably related to edge types, as there is another method with specific edge type https://gephi.org/graphstore/apidocs/org/gephi/graph/api/Graph.html#getEdge-org.gephi.graph.api.Node-org.gephi.graph.api.Node-int- and it seems to be used by the code with `EDGE_TYPE = 1` always, which is incorrect. Edges by default have null type (`0`). Anyway that's just a guess, I would have to review that code.

Also, you should not be using 0.9.2-SNAPSHOT as a dependency, but `0.9.2` which is the release.
But if you want to include latest updates (they don't change anything about graph API), then use next snapshot: `0.9.3-SNAPSHOT`

**Comment by joancf** [comment]
I when I use the 0.9.2 release in the POM's then I can run gephi with the plugin 
And the plugin works, but the gephi update is on the bottom-right part of gephi, if I update gephi then the error appears,after that the code I have problems to debug or if I change any line in the code for debugging , i need to rebuild, reopen, re-install updates and rerun...) 

The call in the code is
Graph.getEdge(Node,Node) without any type, so it should not filter by type?
The code defines  EDGE_TYPE = 1 for creating edges.
If types can be 0(no type?) 1 or 2 ? I can't find the type definition!

PD. I had to change the default pom of the gephi-plugins with  
               <plugin>
                    <groupId>org.codehaus.mojo</groupId>
                    <artifactId>nbm-maven-plugin</artifactId>
                    <version>3.14</version>
                     <extensions>true</extensions> 
                      <inherited>true</inherited>	
                </plugin>

**Comment by joancf** [comment]
I can't use the 0.9.3-SNAPSHOT i get the following error
Failed to execute goal on project gephi-plugins: Could not resolve dependencies for project org.gephi:gephi-plugins:pom:0.9.3-SNAPSHOT: Could not find artifact org.gephi:gephi:tar.gz:linux:0.9.3-SANPSHOT in oss-sonatype (https://oss.sonatype.org/content/repositories/snapshots/) -> [Help 1]

**Comment by eduramiba** [comment]
Edge types have to be added before being used, with https://gephi.org/gephi/0.9.2/apidocs/org/gephi/graph/api/GraphModel.html#addEdgeType-java.lang.Object-

So, the int is just the id returned by that method, when passing any object you want (typically an string such as `TYPE_1`, very similar to neo4j labels). But the int id `1` is to be returned by that method. Also there can be as many edge types as you need.

Looking at https://github.com/gephi/graphstore/blob/edea84830484b545ed631817172f4a411e6b99ef/store/src/main/java/org/gephi/graph/impl/GraphStore.java#L347 and then https://github.com/gephi/graphstore/blob/edea84830484b545ed631817172f4a411e6b99ef/store/src/main/java/org/gephi/graph/impl/EdgeStore.java#L431 it's clear that the method without id, defaults to the NULL_ID. Maybe we should add this to the javadoc of the method, as it's not clear. So, if you have multiple edge types, you have to iterate them all to get all edges for 2 nodes.

Also, you should not need to change the gephi-plugins pom in any way, why do you need it?
Right, for using the 0.9.3 SNAPSHOT more changes are needed, but that's not the idea or necessary at all, you have to develop plugins using 0.9.2, using the poms as they are.

**Comment by eduramiba** [comment]
You might be seeing the problem of hard coded edge type = 1 only after updates since graphml now imports and exports the edge type (label), due to https://github.com/gephi/gephi/issues/1516

So, I guess the bug appears after importing an exported graphml which now includes edge labels, creating edge types during import, and making edge type = 1 not "reserved" anymore to the plugin.

In summary: the plugin is making bad use of edge types, having a potential problem that is starting to happen now with graphml. It probably won't happen if you export to other formats that don't include the edge label.

**Comment by joancf** [comment]
Another issue since last update is that the progess bar does not update its status. So it stays at 0% and at the end it does the full progress at full speed. 
I will commit some changes...just a way  to get the edge type, maybe there are easy ways to get it, but using the iterator, if I only get the first item, then the graph gets locked.

**Comment by eduramiba** [comment]
If you don't use the full iterator, you need to do edgeIterable.doBreak()

**Comment by joancf** [comment]
ok I included this and did pull request. 
As maybe another issue I think that something has changed in the last update of the interface so the progress bar, does not update.

**Comment by eduramiba** [comment]
When does that happen? Make sure the plugin is always doing `Progress.finish` (no early returns, etc), it seems it won't always do it: https://github.com/joancf/gephi-plugins/blob/9a6a24fc2a8338a2c9ff6159e57eb895cde4da26/modules/Multimode-Networks/src/main/java/cz/cvut/fit/gephi/multimode/LongTaskTransformation.java#L361

**Comment by joancf** [comment]
I think it does. The if (online) is to check that the plugin is running as part of the GUI not as library.  But not sure that all the calls are correct.

Progress.start(progressTicket, firstVertical.size()+6);
 several .... Progress.progress (progressTicket,value); 
Progress.finish(progressTicket);

but there are two Progress.start.. maybe this  can be the problem?

**Comment by eduramiba** [comment]
Mmm, right, there should only be one start, as many progress as you want and one finish (better in a finally block).

**Comment by joancf** [comment]
I think it also fails when working over a filtered graph. 
it gives the error 
java.lang.IllegalArgumentException: Edge should belong to a store
I think that the problem is that it tries to add the edge at the visible graph:
graphModel.getGraphVisible();
instead to 
graphModel.getGraph();
if it's that I'll try to solve all these problems and uplad it. 
Also I'm trying to do some optimizations .  (time is cubic, there is nothing to do with this, except working with lists instead of matrices... depending on the graph density)  but for example only upper diagonal results are needed in many cases

**Comment by nelsonmau** [comment]
I had a similar problem while projecting the network: adding an `ID` column to the edges table the problem seems solved and the projection worked well. Never happened before.

**Comment by eduramiba** [comment]
Should be now fixed with https://github.com/gephi/gephi-plugins/pull/173 and https://github.com/eduramiba/gephi-plugins/commit/083b56b04fd1bb40024f0badf206eb981a11af56
